pytensor/tensor/basic.py

```diff
@@ -1461,45 +1494,57 @@ def c_code(self, node, name, inp, out, sub):
 
         code = f"""
             npy_intp shape[{ndim}];
+            int need_new_out;
             """
 
         # Initialize shape
         for i, shp_i in enumerate(inp[1:]):
-            code += """
-                shape[%(i)s] = ((dtype_%(shp_i)s*) PyArray_DATA(%(shp_i)s))[0];
-                """ % dict(
-                i=i, shp_i=shp_i
-            )
+            code += f"""
+                shape[{i}] = ((dtype_{shp_i}*) PyArray_DATA({shp_i}))[0];
+            """
 
-        code += """
-            int need_new_out = (NULL == %(zz)s);
-            for (int i = 0; i < %(ndim)s; i++)
-                need_new_out = (need_new_out
-                                || (PyArray_DIMS(%(zz)s)[i] != shape[i]));
+        # Add checks for runtime broadcasting
+        v_static_shape = node.inputs[0].type.shape
+        o_static_shape = node.outputs[0].type.shape
+        v_ndim = len(v_static_shape)
+        o_ndim = len(o_static_shape)
+        for i, (v_static_dim, out_static_dim) in enumerate(
+            zip(v_static_shape[::-1], o_static_shape[::-1])
+        ):
+            if v_static_dim is None:
+                code += f"""
+                if (PyArray_DIMS({vv})[{v_ndim - i - 1}] == 1 && shape[{o_ndim - i - 1}] != 1)
```

Comment:
Can we be sure that the arrays are long enough for the indices? Unless this is guaranteed for some reason even for invalid inputs I think an explicit check would be good.

Reply:
The docs say this is guaranteed.

https://pytensor.readthedocs.io/en/latest/extending/creating_a_c_op.html#simple-cop-example

> Also, in the C code, it is very important to properly validate the inputs
and outputs storage. PyTensor guarantees that the inputs exist and have the
right number of dimensions but it does not guarantee their exact shape. For
instance, if an :class:`Op` computes the sum of two vectors, it needs to validate that
its two inputs have the same shape. In our case, we do not need to validate
the exact shapes of the inputs because we don't have a need that they match
in any way.

> For the outputs, things are a little bit more subtle. PyTensor does not
guarantee that they have been allocated but it does guarantee that, if they
have been allocated, they have the right number of dimension. Again, PyTensor
offers no guarantee on the exact shapes. This means that, in our example, we
need to validate that the output storage has been allocated and has the same
shape as our vector input. If it is not the case, we allocate a new output
storage with the right shape and number of dimensions.

This is not true if `fn.trust_input=True`, and/or if an Op returns an output with the wrong shape, so I don't know what they mean by "guarantee". However this probably means most Ops are written with this assumption?